### PR TITLE
feat(runtime): Expose epoch height to WASM runtime

### PR DESCRIPTION
**Pull request by ilblackdragon**
Protocol change: adding new WASM API `epoch_height`

Implements https://github.com/nearprotocol/NEPs/pull/47

Test Plan
----
All tests pass
External smart contracts can use epoch_height to determine current epoch's ordinal.

### CODE REVIEW

**Comment by ilblackdragon** [comment]
Going to split into 3 PRs.

File: runtime/runtime/src/actions.rs [hunk 1]
```diff
@@ -162,6 +162,7 @@ pub(crate) fn action_function_call(
         input: function_call.args.clone(),
         block_index: apply_state.block_index,
         block_timestamp: apply_state.block_timestamp,
+        epoch_id: ((apply_state.epoch_id.0).0).0,
```
Comment (frol) [inline runtime/runtime/src/actions.rs]: O_o.o.o

Can we avoid this nesting (implement `.into()` here)?

File: runtime/runtime/src/state_viewer.rs [hunk 2]
```diff
@@ -115,6 +117,7 @@ impl TrieViewer {
                 input: args.to_owned(),
                 block_index: block_height,
                 block_timestamp,
+                epoch_id: ((epoch_id.0).0).0,
```
Comment (frol) [inline runtime/runtime/src/state_viewer.rs]: Another .0.0.0

File: runtime/runtime/src/state_viewer.rs [hunk 3]
```diff
@@ -180,7 +183,16 @@ mod tests {
         let (viewer, root) = get_test_trie_viewer();
 
         let mut logs = vec![];
-        let result = viewer.call_function(root, 1, 1, &alice_account(), "run_test", &[], &mut logs);
+        let result = viewer.call_function(
+            root,
+            1,
+            1,
+            &Default::default(),
```
Comment (frol) [inline runtime/runtime/src/state_viewer.rs]: It is already not great, but we can at least use `&EpochId::default()`.

**Comment by evgenykuzyakov** [comment]
Global question:
We don't expose `block_hash` to the Runtime contracts. We only expose `block_height`. It's unclear the benefits of exposing the full hash to the contracts instead of just the number, or something like this.
Solutions:
1) Expose `u64` of the block_height when the Epoch switched.
2) Expose `u64` which is the actual height (number) of the epoch starting from the genesis. This seems to require additional on top of `EpochManager` to be able to get this number.

Minor:
We don't need to pass full `EpochId` to the `ApplyState` if the Runtime only uses the CryptoHash inside. It would make sense to pass just the CryptoHash from the EpochId.

**Comment by evgenykuzyakov** [comment]
#2062

**Comment by ilblackdragon** [comment]
@bowenwang1996 @SkidanovAlex What do you think to have an ordinal for an epoch, given we can not have two competing epochs now with NFG? I think indeed it will be more useful than the current EpochId.

We also probably broke abstraction somewhere, but originally idea behind EpochId type is that we can change it to u64 and it would work.

**Comment by bowenwang1996** [comment]
It should work, but I need to examine the code more carefully before fully confirming this.

File: runtime/near-vm-runner/src/imports.rs [hunk 4]
```diff
@@ -55,6 +55,7 @@ wrapped_imports! {
     // TODO #1903 rename to `block_height`
```
Comment (frol) [inline runtime/near-vm-runner/src/imports.rs]: It is unrelated to this PR, but given we break it here, should we also address this?

Comment (ilblackdragon) [reply runtime/near-vm-runner/src/imports.rs]: We don't break here anything. This is additive change not requiring any changes on existing contracts. Switching to different function name would require to recompile all deployed contracts from my understanding @nearmax

**Comment by ilblackdragon** [comment]
Addressed @evgenykuzyakov comment - expose epoch_height as ordinal of epoch since genesis now instead of not very useful hash.

File: near/src/runtime.rs [hunk 5]
```diff
@@ -286,9 +295,12 @@ impl NightshadeRuntime {
             }
         };
 
+        let epoch_height = self.get_epoch_height_from_prev_block(prev_block_hash)?;
```
Comment (nearmax) [inline near/src/runtime.rs]: The epoch height computed from the previous block might not be the same as the epoch height of the current block.

File: runtime/near-vm-runner/src/imports.rs (hunk 4 above)
Comment (nearmax) [reply runtime/near-vm-runner/src/imports.rs]: Yes, renaming the host function is harder and so we postpone it for the future. We probably will do a soft deprecation with warning and keep the old name forever.

Comment (frol) [reply runtime/near-vm-runner/src/imports.rs]: Do we want to add `block_height` field here or we don’t want to duplicate value for consistency reasons only and should just work around this on SDK level?

In any case, the comment and the issue will benefit if it gets updated with the current decision.

I’m sorry to hijack the PR discussion.

Comment (nearmax) [reply runtime/near-vm-runner/src/imports.rs]: Sure, let's add it here.

File: near/src/runtime.rs (hunk 5 above)
Comment (ilblackdragon) [reply near/src/runtime.rs]: epoch height from prev block meaning that given prev block hash we look up epoch id for next block. That's why it's not `get_epoch_height(prev_block_hash)` but `get_epoch_heiight_from_prev_block`. This is how epoch manager normally operated (see other get_..._from_prev_block()` functions.

File: core/primitives/src/types.rs [hunk 6]
```diff
@@ -24,8 +24,10 @@ pub type StorageUsage = u64;
 pub type StorageUsageChange = i64;
 /// Nonce for transactions.
 pub type Nonce = u64;
-/// Index of the block.
+/// Height of the block from genesis.
```
Comment (frol) [inline core/primitives/src/types.rs]: Well, "from genesis" is not completely accurate, as we preserve the block height on hard forks while the genesis gets changed.

File: chain/epoch_manager/src/lib.rs [hunk 7]
```diff
@@ -917,7 +917,7 @@ impl EpochManager {
         Ok(EpochId(first_block_info.prev_hash))
     }
 
-    fn get_epoch_info(&mut self, epoch_id: &EpochId) -> Result<&EpochInfo, EpochError> {
+    pub fn get_epoch_info(&mut self, epoch_id: &EpochId) -> Result<&EpochInfo, EpochError> {
```
Comment (bowenwang1996) [inline chain/epoch_manager/src/lib.rs]: I thought the idea is to not make this public. If we do then we can get rid of a lot of method in epoch manager can directly implement them in runtime adapter.

File: runtime/near-vm-runner/src/imports.rs (hunk 4 above)
Comment (nearmax) [reply runtime/near-vm-runner/src/imports.rs]: @ilblackdragon ^

**Comment by nearmax** [comment]
Please also add it near-sdk-rs

**Comment by ilblackdragon** [comment]
@nearmax -> that TODO will stay TODO from my understanding. There is no point of breaking all of our contract for renaming.

@nearmax There is an old PR https://github.com/near/near-sdk-rs/pull/71 in near-sdk-rs, but want to submit this first so there is a node to run against.

File: chain/epoch_manager/src/lib.rs (hunk 7 above)
Comment (ilblackdragon) [reply chain/epoch_manager/src/lib.rs]: Might be. I couldn't find a reason now to not expose it. We can clean up runtime adapter separately.

**Comment by frol** [comment]
@ilblackdragon 
> that TODO will stay TODO from my understanding. There is no point of breaking all of our contract for renaming.

The point is that the TODO is not up-to-date with the decision we currently have: there is no way we rename/remove that field since that will require to recompile every single smart contract.

File: runtime/near-vm-runner/src/imports.rs (hunk 4 above)
Comment (ilblackdragon) [reply runtime/near-vm-runner/src/imports.rs]: I added details where they belong - https://github.com/nearprotocol/nearcore/issues/1903

If someone wants to actually add `block_height` - that indeed sounds like a separate PR.
Note also it should start with spec change.

**Pull request merged**